metaseq/tasks/streaming_finetune_language_modeling.py

```diff
@@ -0,0 +1,147 @@
+        )
+        src_tokens_len = len(self.tokenizer.encode(src).ids)
+        tgt_tokens = torch.clone(full_tokens)
+        tgt_tokens[:src_tokens_len] = self.dictionary.pad_index
+        return (full_tokens, tgt_tokens)
+
+    def load_dataset(self, split: str, epoch=1, combine=False, **kwargs):
+        """Load a given dataset split.
+
+        The folder structure is assumed to look like:
+
+            /path/to/data/train/00/foo.jsonl
+            /path/to/data/train/00/bar.jsonl
+            /path/to/data/train/01/foo.jsonl
+            /path/to/data/train/01/bar.jsonl
+            /path/to/data/valid/00/foo.jsonl
+            /path/to/data/valid/00/bar.jsonl
+
+        In this example, we have two "shards" of training data, which will be
+        iterated over in epochs 1 and 2, respectively. Subsequent epochs will
+        cycle back over the same data. We also have two different data sources
+        in each shard (foo and bar), which will be combined and shuffled.
+
+        Args:
+            split (str): name of the split (e.g., train, valid, valid1, test)
+        """
+        # This function reads a bunch of jsonl files, concats them together,
+        # shuffles them, then chunks them into blocks of tokens (e.g., 2048).
+
+        # determine number of shards for this split shards = {}
+        cur_shard_str = self.get_shard_str(epoch, split)
+
+        # concatenate any jsonl files that are part of the shard
+        datasets, corpora = [], []
+        for file in sorted(
+            os.listdir(os.path.join(self.args.data, split, cur_shard_str))
+        ):
+            if not file.endswith(".jsonl"):
+                continue
+            datasets.append(
+                JsonlDataset(
+                    path=os.path.join(self.args.data, split, cur_shard_str, file),
+                    tokenizer=self._tokenize_src_tgt_json,
+                )
+            )
+            corpora.append(os.path.splitext(file)[0])
+        assert len(datasets) > 0
+
+        if self.args.multicorpus_sampling_alpha != 1:
+            raise NotImplementedError
+
+        dataset = torch.utils.data.ConcatDataset(datasets)
+
+        # shuffle order across epochs
+        dataset = StreamingShuffleDataset(dataset, seed=self.args.seed)
+
+        self.datasets[split] = StreamingSrcTgtDataset(
+            dataset,
+            # We generate blocks with one extra token, so that we have a target
+            # for the final input token. This results in slight data loss.
+            block_size=self.args.tokens_per_sample + 1,
+            break_mode=self.args.sample_break_mode,
+            # we drop the remainder block during training
+            drop_last=(split == "train"),
+            padding_idx=self.source_dictionary.pad(),
+            # 1284 is a randomly-generated offset to decouple the seed used here
+            # from the seed used above in StreamingShuffleDataset
+            seed=1284 + self.args.seed,
```

Comment:
Nit: Would be good to keep track of all the RNG offsets in our codebase and where collisions may occur (or add a TODO pointing to https://github.com/facebookresearch/metaseq/issues/65)

Reply:
Will add TODO for now.